File: python/line_solver/inference/api/infer_lqn.py
# Copyright (c) 2012-2026, Imperial College London
# All rights reserved.
"""LQN parameter identification via an Extended Kalman Filter.

Estimates hidden Layered Queueing Network parameters (activity host demands,
task think times) from measured performance data (response times, utilizations,
throughputs), following Zheng, Yang, Woodside, Litoiu, Iszlai, "Tracking
Time-Varying Parameters in Software Systems with Extended Kalman Filters",
CASCON 2005 (equations 1-9).

The parameter is modelled as a zero-mean random walk a_k = a_{k-1} + w and the
measurement as z_k = h(a_k) + v, where h is the LQN performance model. This is
the native-Python counterpart of the MATLAB ``infer_lqn`` family and the JAR
``jline.api.infer.InferLqn``; results match across codebases.

A parameter is specified as a dict {'type': 'hostdem'|'think', 'name': <name>};
an observation as {'metric': 'RespT'|'Util'|'Tput'|'QLen', 'name': <name>}.
"""
import numpy as np

__all__ = [
    'infer_lqn',
    'infer_lqn_ekf',
    'infer_lqn_jacobian',
    'infer_lqn_setparams',
    'infer_lqn_getparams',
    'infer_lqn_getobs',
    'infer_lqn_findbyname',
]


def infer_lqn_findbyname(container, name):
    """Return the first element of container whose .name equals name, else None."""
    for el in container:
        if getattr(el, 'name', None) == name:
            return el
    return None


def infer_lqn_setparams(model, param_spec, a):
    """Apply a parameter vector to a LayeredNetwork in place.

    param_spec[i] is a dict with keys 'type' ('hostdem' | 'think') and 'name'
    (activity name for 'hostdem', task name for 'think'). Means are injected as
    exponential distributions (SCV = 1).
    """
    a = np.asarray(a, dtype=float).ravel()
    if a.size != len(param_spec):
        raise ValueError('Length of parameter vector does not match param_spec.')
    for i, spec in enumerate(param_spec):
        val = float(a[i])
        typ = str(spec['type']).lower()
        if typ == 'hostdem':
            act = infer_lqn_findbyname(model.activities, spec['name'])
            if act is None:
                raise ValueError("Activity '%s' not found." % spec['name'])
            act.setHostDemand(val)
        elif typ == 'think':
            tsk = infer_lqn_findbyname(model.tasks, spec['name'])
            if tsk is None:
                raise ValueError("Task '%s' not found." % spec['name'])
            tsk.setThinkTime(val)
        else:
            raise ValueError("Unknown parameter type '%s'." % spec['type'])
    return model


def infer_lqn_getparams(model, param_spec):
    """Read the current values of the parameters named in param_spec."""
    a0 = np.zeros(len(param_spec))
    for i, spec in enumerate(param_spec):
        typ = str(spec['type']).lower()
        if typ == 'hostdem':
            act = infer_lqn_findbyname(model.activities, spec['name'])
            if act is None:
                raise ValueError("Activity '%s' not found." % spec['name'])
            a0[i] = act.getHostDemandMean()
        elif typ == 'think':
            tsk = infer_lqn_findbyname(model.tasks, spec['name'])
            if tsk is None:
                raise ValueError("Task '%s' not found." % spec['name'])
            a0[i] = tsk.getThinkTimeMean()
        else:
            raise ValueError("Unknown parameter type '%s'." % spec['type'])
    return a0


def infer_lqn_getobs(names, metrics, obs_spec):
    """Extract the observation vector selected by obs_spec.

    names is the element name list aligned with the metric vectors (as in
    LayeredNetworkStruct.names). metrics is a dict with keys QLen/Util/RespT/Tput
    (each a vector indexed by element). obs_spec[i] is a dict with keys 'metric'
    ('RespT'|'Util'|'Tput'|'QLen') and 'name' (element name).
    """
    names_list = list(names)
    z = np.zeros(len(obs_spec))
    for i, spec in enumerate(obs_spec):
        idx = None
        for k, nm in enumerate(names_list):
            if nm == spec['name']:
                idx = k
                break
        if idx is None:
            raise ValueError("Element '%s' not found in the LQN." % spec['name'])
        metric = str(spec['metric']).lower()
        if metric == 'respt':
            z[i] = metrics['RespT'][idx]
        elif metric == 'util':
            z[i] = metrics['Util'][idx]
        elif metric == 'tput':
            z[i] = metrics['Tput'][idx]
        elif metric == 'qlen':
            z[i] = metrics['QLen'][idx]
        else:
            raise ValueError("Unknown metric '%s'." % spec['metric'])
    return z


def infer_lqn_jacobian(hfun, a, fd_step=1e-3, fd_floor=1e-6):
    """Forward finite-difference sensitivity matrix H = dh/da and h0 = hfun(a).

    Column i is (hfun(a + d_i) - h0) / d_i with d_i = fd_step*max(|a_i|, fd_floor).
    This is the approximate sensitivity matrix H_k used in the EKF update.
    """
    a = np.asarray(a, dtype=float).ravel()
    h0 = np.asarray(hfun(a), dtype=float).ravel()
    H = np.zeros((h0.size, a.size))
    for i in range(a.size):
        d = fd_step * max(abs(a[i]), fd_floor)
        ap = a.copy()
        ap[i] += d
        hi = np.asarray(hfun(ap), dtype=float).ravel()
        H[:, i] = (hi - h0) / d
    return H, h0


def infer_lqn_ekf(hfun, a0, P0, Z, Q, R, options=None):
    """Extended Kalman Filter tracking a hidden parameter vector across Z.

    hfun maps a parameter vector to a predicted observation vector z = h(a).
    a0 (np,), P0 (np,np), Z (no,nsteps), Q (np,np), R (no,no). Returns
    (ahat (np,nsteps), info) where info has keys P, Phist, e, zpred, Er and Ea
    (parameter tracking RMS vs options['aTrue'] when supplied, else None).
    """
    if options is None:
        options = {}
    fd_step = options.get('fdStep', 1e-3)
    fd_floor = options.get('fdFloor', 1e-6)
    clamp = options.get('clampPositive', True)
    a_true = options.get('aTrue', None)
    verbose = options.get('verbose', False)

    a0 = np.asarray(a0, dtype=float).ravel()
    Z = np.asarray(Z, dtype=float)
    if Z.ndim == 1:
        Z = Z.reshape(-1, 1)
    no, nsteps = Z.shape
    npar = a0.size

    ahat = np.zeros((npar, nsteps))
    e_hist = np.zeros((no, nsteps))
    zpred_hist = np.zeros((no, nsteps))
    Phist = []

    a = a0.copy()
    P = np.asarray(P0, dtype=float).copy()
    I_np = np.eye(npar)
    for k in range(nsteps):
        # (1) predict: zero-mean drift; project covariance (eq 5)
        a_pred = a
        Ppred = P + Q
        # (2,4) predicted measurement and sensitivity matrix
        H, zpred = infer_lqn_jacobian(hfun, a_pred, fd_step, fd_floor)
        # (3) prediction error
        e = Z[:, k] - zpred
        # (6) Kalman gain (suboptimal because h is nonlinear)
        S = H @ Ppred @ H.T + R
        K = Ppred @ H.T @ np.linalg.inv(S)
        # (4-update) improved estimate
        a = a_pred + K @ e
        if clamp:
            a = np.maximum(a, fd_floor)
        # (7) covariance update; symmetrize for stability
        P = (I_np - K @ H) @ Ppred
        P = 0.5 * (P + P.T)

        ahat[:, k] = a
        e_hist[:, k] = e
        zpred_hist[:, k] = zpred
        Phist.append(P.copy())
        if verbose:
            print('[infer_lqn_ekf] step %d/%d  ||e||=%.4g' % (k + 1, nsteps, np.linalg.norm(e)))

    info = {'P': P, 'Phist': Phist, 'e': e_hist, 'zpred': zpred_hist}
    info['Er'] = float(np.sqrt(np.mean(e_hist ** 2)))
    if a_true is not None:
        a_true = np.asarray(a_true, dtype=float).ravel()
        D = ahat - a_true.reshape(-1, 1)
        info['Ea'] = float(np.sqrt(np.mean(D ** 2)))
    else:
        info['Ea'] = None
    return ahat, info


def infer_lqn(model, param_spec, obs_spec, Z, options=None):
    """Identify hidden LQN parameters from measured performance data.

    Estimates the parameters named in param_spec (activity host demands and/or
    task think times) of the LayeredNetwork model from the measurement sequence
    Z (no x nsteps) using an EKF over the observation model defined by obs_spec.
    A single measurement column with options['QFac'] = 0 reduces to one-shot
    least-squares calibration.

    options keys (defaults): solver (callable model->solver, default native
    SolverLN), QFac (0.1), RFac (0.2), cvA (1), gammaT ([]; else T/Tstar; else 1),
    T, Tstar, Q, R, P0, a0, aTrue, fdStep (1e-3), fdFloor (1e-6),
    clampPositive (True), verbose (False).

    Returns (model, info): model has the final estimate applied; info is the EKF
    result dict augmented with ahat, a0, Q, R, P0.
    """
    if options is None:
        options = {}
    eps = np.finfo(float).eps

    no = len(obs_spec)
    Z = np.asarray(Z, dtype=float)
    if Z.ndim == 1:
        Z = Z.reshape(-1, 1)
    if Z.shape[0] != no:
        raise ValueError('Row count of Z must equal len(obs_spec).')

    solver_ctor = options.get('solver', None)

    a0 = options.get('a0', None)
    if a0 is None:
        a0 = infer_lqn_getparams(model, param_spec)
    a0 = np.asarray(a0, dtype=float).ravel()

    QFac = options.get('QFac', 0.1)
    RFac = options.get('RFac', 0.2)
    cvA = options.get('cvA', 1.0)
    gammaT = options.get('gammaT', None)
    if gammaT is None:
        T = options.get('T', None)
        Tstar = options.get('Tstar', None)
        gammaT = (T / Tstar) if (T is not None and Tstar is not None) else 1.0

    Q = options.get('Q', None)
    if Q is None:
        qd = (QFac * np.abs(a0) * cvA) ** 2                 # eq 9a
        Q = np.diag(np.maximum(qd, eps))
    R = options.get('R', None)
    if R is None:
        zbar = np.mean(Z, axis=1)
        rd = ((RFac * np.abs(zbar)) / 1.96) ** 2 / gammaT   # eq 9b
        R = np.diag(np.maximum(rd, eps))
    P0 = options.get('P0', None)
    if P0 is None:
        P0 = np.diag(np.maximum((0.5 * np.abs(a0)) ** 2, eps))

    def hfun(a):
        infer_lqn_setparams(model, param_spec, a)
        if solver_ctor is not None:
            solver = solver_ctor(model)
        else:
            from line_solver import SolverLN
            solver = SolverLN(model, verbose=False)
        QN, UN, RN, TN, _AN, _WN = solver.get_ensemble_avg()
        names = solver.lqn.names
        metrics = {'QLen': QN, 'Util': UN, 'RespT': RN, 'Tput': TN}
        return infer_lqn_getobs(names, metrics, obs_spec)

    ekfopts = {
        'fdStep': options.get('fdStep', 1e-3),
        'fdFloor': options.get('fdFloor', 1e-6),
        'clampPositive': options.get('clampPositive', True),
        'aTrue': options.get('aTrue', None),
        'verbose': options.get('verbose', False),
    }
    ahat, info = infer_lqn_ekf(hfun, a0, P0, Z, Q, R, ekfopts)
    info['ahat'] = ahat
    info['a0'] = a0
    info['Q'] = Q
    info['R'] = R
    info['P0'] = P0

    infer_lqn_setparams(model, param_spec, ahat[:, -1])
    return model, info

File: jar/src/main/java/jline/api/infer/InferLqn.java
package jline.api.infer;

import jline.VerboseLevel;
import jline.lang.layered.Activity;
import jline.lang.layered.LayeredNetwork;
import jline.lang.layered.Task;
import jline.solvers.LayeredNetworkAvgTable;
import jline.solvers.SolverOptions;
import jline.solvers.ln.LNOptions;
import jline.solvers.ln.SolverLN;
import jline.util.matrix.Matrix;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

/**
 * LQN parameter identification via an Extended Kalman Filter.
 *
 * <p>Estimates hidden Layered Queueing Network parameters (activity host
 * demands, task think times) from measured performance data (response times,
 * utilizations, throughputs), following Zheng, Yang, Woodside, Litoiu, Iszlai,
 * "Tracking Time-Varying Parameters in Software Systems with Extended Kalman
 * Filters", CASCON 2005 (equations 1-9). The parameter is modelled as a
 * zero-mean random walk a_k = a_{k-1} + w and the measurement as z_k = h(a_k)+v,
 * where h is the LQN performance model. This is the JAR counterpart of the
 * MATLAB {@code infer_lqn} family and the native-Python
 * {@code line_solver.inference.infer_lqn}; results match across codebases.</p>
 *
 * Copyright (c) 2012-2026, Imperial College London. All rights reserved.
 */
public class InferLqn {

    private static final double EPS = 2.220446049250313e-16;

    private InferLqn() {
    }

    /** Finite-difference sensitivity matrix H = dh/da together with h0 = h(a). */
    public static class JacobianResult {
        public final Matrix H;
        public final Matrix h0;

        public JacobianResult(Matrix H, Matrix h0) {
            this.H = H;
            this.h0 = h0;
        }
    }

    /**
     * Apply a parameter vector to a LayeredNetwork in place.
     *
     * @param model     the LQN whose parameters are set
     * @param paramSpec list of parameters to set
     * @param a         (np x 1) parameter vector
     */
    public static void setParams(LayeredNetwork model, List<ParamSpec> paramSpec, Matrix a) {
        if (a.getNumRows() * a.getNumCols() != paramSpec.size()) {
            throw new RuntimeException("Length of parameter vector does not match paramSpec.");
        }
        for (int i = 0; i < paramSpec.size(); i++) {
            ParamSpec s = paramSpec.get(i);
            double val = a.get(i);
            switch (s.type) {
                case HOSTDEM: {
                    Activity act = findActivity(model, s.name);
                    if (act == null) {
                        throw new RuntimeException("Activity '" + s.name + "' not found.");
                    }
                    act.setHostDemand(val);
                    break;
                }
                case THINK: {
                    Task tsk = findTask(model, s.name);
                    if (tsk == null) {
                        throw new RuntimeException("Task '" + s.name + "' not found.");
                    }
                    tsk.setThinkTime(val);
                    break;
                }
                default:
                    throw new RuntimeException("Unknown parameter type.");
            }
        }
        // invalidate and regenerate the cached layered struct so the next solve
        // re-reads the mutated model objects
        model.getStruct(true);
    }

    /** Read the current values of the parameters named in paramSpec. */
    public static Matrix getParams(LayeredNetwork model, List<ParamSpec> paramSpec) {
        Matrix a0 = new Matrix(paramSpec.size(), 1);
        for (int i = 0; i < paramSpec.size(); i++) {
            ParamSpec s = paramSpec.get(i);
            switch (s.type) {
                case HOSTDEM: {
                    Activity act = findActivity(model, s.name);
                    if (act == null) {
                        throw new RuntimeException("Activity '" + s.name + "' not found.");
                    }
                    a0.set(i, 0, act.getHostDemandMean());
                    break;
                }
                case THINK: {
                    Task tsk = findTask(model, s.name);
                    if (tsk == null) {
                        throw new RuntimeException("Task '" + s.name + "' not found.");
                    }
                    a0.set(i, 0, tsk.getThinkTimeMean());
                    break;
                }
                default:
                    throw new RuntimeException("Unknown parameter type.");
            }
        }
        return a0;
    }

    /** Extract the observation vector selected by obsSpec from a solved table. */
    public static Matrix getObs(LayeredNetworkAvgTable table, List<ObsSpec> obsSpec) {
        List<String> names = table.getNodeNames();
        List<Double> qlen = table.getQLen();
        List<Double> util = table.getUtil();
        List<Double> respt = table.getRespT();
        List<Double> tput = table.getTput();
        Matrix z = new Matrix(obsSpec.size(), 1);
        for (int i = 0; i < obsSpec.size(); i++) {
            ObsSpec s = obsSpec.get(i);
            int idx = names.indexOf(s.name);
            if (idx < 0) {
                throw new RuntimeException("Element '" + s.name + "' not found in the LQN.");
            }
            double v;
            switch (s.metric) {
                case RESPT:
                    v = respt.get(idx);
                    break;
                case UTIL:
                    v = util.get(idx);
                    break;
                case TPUT:
                    v = tput.get(idx);
                    break;
                case QLEN:
                    v = qlen.get(idx);
                    break;
                default:
                    throw new RuntimeException("Unknown metric.");
            }
            z.set(i, 0, v);
        }
        return z;
    }

    /**
     * Forward finite-difference sensitivity matrix of h at a.
     *
     * <p>Column i is (h(a + d_i) - h0)/d_i with d_i = fdStep*max(|a_i|, fdFloor).
     * hfun is evaluated np+1 times. This is the approximate sensitivity matrix
     * H_k used in the EKF update.</p>
     */
    public static JacobianResult jacobian(Function<Matrix, Matrix> hfun, Matrix a,
                                          double fdStep, double fdFloor) {
        int np = a.getNumRows();
        Matrix h0 = hfun.apply(a);
        int no = h0.getNumRows();
        Matrix H = new Matrix(no, np);
        for (int i = 0; i < np; i++) {
            double d = fdStep * Math.max(Math.abs(a.get(i, 0)), fdFloor);
            Matrix ap = a.copy();
            ap.set(i, 0, ap.get(i, 0) + d);
            Matrix hi = hfun.apply(ap);
            for (int r = 0; r < no; r++) {
                H.set(r, i, (hi.get(r, 0) - h0.get(r, 0)) / d);
            }
        }
        return new JacobianResult(H, h0);
    }

    /**
     * Extended Kalman Filter tracking a hidden parameter vector across Z.
     *
     * @param hfun maps a parameter vector to a predicted observation z = h(a)
     * @param a0   (np x 1) initial estimate
     * @param P0   (np x np) initial covariance
     * @param Z    (no x nsteps) measurement matrix, one column per step
     * @param Q    (np x np) drift covariance
     * @param R    (no x no) measurement covariance
     * @param opt  options (fdStep, fdFloor, clampPositive, aTrue, verbose)
     */
    public static InferLqnResult ekf(Function<Matrix, Matrix> hfun, Matrix a0, Matrix P0,
                                     Matrix Z, Matrix Q, Matrix R, InferLqnOptions opt) {
        if (opt == null) {
            opt = new InferLqnOptions();
        }
        int np = a0.getNumRows();
        int no = Z.getNumRows();
        int nsteps = Z.getNumCols();

        Matrix ahat = new Matrix(np, nsteps);
        Matrix eHist = new Matrix(no, nsteps);
        Matrix zpredHist = new Matrix(no, nsteps);
        List<Matrix> Phist = new ArrayList<Matrix>();

        Matrix a = a0.copy();
        Matrix P = P0.copy();
        Matrix I = Matrix.eye(np);

        for (int k = 0; k < nsteps; k++) {
            // (1) predict: zero-mean drift; project covariance (eq 5)
            Matrix aPred = a;
            Matrix Ppred = P.add(Q);
            // (2,4) predicted measurement and sensitivity matrix
            JacobianResult jr = jacobian(hfun, aPred, opt.fdStep, opt.fdFloor);
            Matrix H = jr.H;
            Matrix zpred = jr.h0;
            // (3) prediction error
            Matrix zk = new Matrix(no, 1);
            for (int r = 0; r < no; r++) {
                zk.set(r, 0, Z.get(r, k));
            }
            Matrix e = zk.sub(zpred);
            // (6) Kalman gain (suboptimal because h is nonlinear)
            Matrix S = H.mult(Ppred).mult(H.transpose()).add(R);
            Matrix K = Ppred.mult(H.transpose()).mult(S.inv());
            // (4-update) improved estimate
            a = aPred.add(K.mult(e));
            if (opt.clampPositive) {
                for (int r = 0; r < np; r++) {
                    if (a.get(r, 0) < opt.fdFloor) {
                        a.set(r, 0, opt.fdFloor);
                    }
                }
            }
            // (7) covariance update; symmetrize for numerical stability
            P = I.sub(K.mult(H)).mult(Ppred);
            P = P.add(P.transpose()).scale(0.5);

            for (int r = 0; r < np; r++) {
                ahat.set(r, k, a.get(r, 0));
            }
            for (int r = 0; r < no; r++) {
                eHist.set(r, k, e.get(r, 0));
                zpredHist.set(r, k, zpred.get(r, 0));
            }
            Phist.add(P.copy());
            if (opt.verbose) {
                System.out.printf("[InferLqn.ekf] step %d/%d  ||e||=%.4g%n", k + 1, nsteps, norm(e));
            }
        }

        InferLqnResult res = new InferLqnResult();
        res.ahat = ahat;
        res.e = eHist;
        res.zpred = zpredHist;
        res.P = P;
        res.Phist = Phist;
        res.Er = rms(eHist);
        if (opt.aTrue != null) {
            double acc = 0.0;
            int cnt = 0;
            for (int r = 0; r < np; r++) {
                double at = opt.aTrue.get(r, 0);
                for (int k = 0; k < nsteps; k++) {
                    double d = ahat.get(r, k) - at;
                    acc += d * d;
                    cnt++;
                }
            }
            res.Ea = Math.sqrt(acc / cnt);
        } else {
            res.Ea = null;
        }
        return res;
    }

    /**
     * Identify hidden LQN parameters from measured performance data.
     *
     * <p>Estimates the parameters named in paramSpec (activity host demands
     * and/or task think times) of the LayeredNetwork model from the measurement
     * sequence Z (no x nsteps) using an EKF over the observation model defined by
     * obsSpec. A single measurement column with opt.QFac = 0 reduces to one-shot
     * least-squares calibration. On return the model has the final estimate
     * applied.</p>
     */
    public static InferLqnResult inferLqn(final LayeredNetwork model, final List<ParamSpec> paramSpec,
                                          final List<ObsSpec> obsSpec, Matrix Z, InferLqnOptions opt) {
        if (opt == null) {
            opt = new InferLqnOptions();
        }
        int no = obsSpec.size();
        if (Z.getNumRows() != no) {
            throw new RuntimeException("Row count of Z must equal obsSpec size.");
        }

        Matrix a0 = (opt.a0 != null) ? opt.a0.copy() : getParams(model, paramSpec);
        int np = a0.getNumRows();

        double gammaT;
        if (opt.gammaT != null) {
            gammaT = opt.gammaT;
        } else if (opt.T != null && opt.Tstar != null) {
            gammaT = opt.T / opt.Tstar;
        } else {
            gammaT = 1.0;
        }

        Matrix Q = (opt.Q != null) ? opt.Q : buildQ(a0, opt.QFac, opt.cvA);   // eq 9a
        Matrix R = (opt.R != null) ? opt.R : buildR(Z, opt.RFac, gammaT);     // eq 9b
        Matrix P0 = (opt.P0 != null) ? opt.P0 : buildP0(a0);

        final InferLqnOptions.ObservationSolver solver = opt.solver;
        Function<Matrix, Matrix> hfun = new Function<Matrix, Matrix>() {
            @Override
            public Matrix apply(Matrix a) {
                setParams(model, paramSpec, a);
                LayeredNetworkAvgTable table;
                if (solver != null) {
                    table = solver.evaluate(model);
                } else {
                    SolverOptions so = new LNOptions();
                    so.verbose = VerboseLevel.SILENT;
                    table = (LayeredNetworkAvgTable) new SolverLN(model, so).getEnsembleAvg();
                }
                return getObs(table, obsSpec);
            }
        };

        InferLqnResult res = ekf(hfun, a0, P0, Z, Q, R, opt);
        res.a0 = a0;
        res.Q = Q;
        res.R = R;
        res.P0 = P0;

        // apply the final estimate to the model
        Matrix aLast = new Matrix(np, 1);
        int nsteps = res.ahat.getNumCols();
        for (int r = 0; r < np; r++) {
            aLast.set(r, 0, res.ahat.get(r, nsteps - 1));
        }
        setParams(model, paramSpec, aLast);
        return res;
    }

    // ---------------- helpers ----------------

    private static Activity findActivity(LayeredNetwork model, String name) {
        for (Activity a : model.getActivities().values()) {
            if (a.getName().equals(name)) {
                return a;
            }
        }
        return null;
    }

    private static Task findTask(LayeredNetwork model, String name) {
        for (Task t : model.getTasks().values()) {
            if (t.getName().equals(name)) {
                return t;
            }
        }
        return null;
    }

    private static Matrix buildQ(Matrix a0, double QFac, double cvA) {
        int np = a0.getNumRows();
        double[] d = new double[np];
        for (int i = 0; i < np; i++) {
            double q = QFac * Math.abs(a0.get(i, 0)) * cvA;
            d[i] = Math.max(q * q, EPS);
        }
        return Matrix.diag(d);
    }

    private static Matrix buildR(Matrix Z, double RFac, double gammaT) {
        int no = Z.getNumRows();
        int nsteps = Z.getNumCols();
        double[] d = new double[no];
        for (int i = 0; i < no; i++) {
            double zbar = 0.0;
            for (int k = 0; k < nsteps; k++) {
                zbar += Z.get(i, k);
            }
            zbar /= nsteps;
            double r = (RFac * Math.abs(zbar)) / 1.96;
            d[i] = Math.max((r * r) / gammaT, EPS);
        }
        return Matrix.diag(d);
    }

    private static Matrix buildP0(Matrix a0) {
        int np = a0.getNumRows();
        double[] d = new double[np];
        for (int i = 0; i < np; i++) {
            double p = 0.5 * Math.abs(a0.get(i, 0));
            d[i] = Math.max(p * p, EPS);
        }
        return Matrix.diag(d);
    }

    private static double norm(Matrix v) {
        double s = 0.0;
        int n = v.getNumRows() * v.getNumCols();
        for (int i = 0; i < n; i++) {
            double x = v.get(i);
            s += x * x;
        }
        return Math.sqrt(s);
    }

    private static double rms(Matrix M) {
        int rows = M.getNumRows();
        int cols = M.getNumCols();
        double s = 0.0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                double x = M.get(r, c);
                s += x * x;
            }
        }
        return Math.sqrt(s / (rows * cols));
    }
}
